cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +363,139 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Upgrades an existing application by upgrading application artifact versions and plugin artifact versions.
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @throws IllegalStateException if something unexpected happened during upgrade.
+   * @throws IOException if there was an IO error during initializing application class from artifact.
+   * @throws JsonIOException if there was an error in serializing or deserializing app config.
+   * @throws Exception if there was an exception during the upgrade of application. This exception will often wrap
```

Comment:
The reason why I have designed it like this is due to the case that application upgrade can fail due to variety of reason which is known only to the function, not the callee (we can implement exception for each of them but that seems unnecessary). Common errors like update not supported for this app and uncommon errors like (no app artifact found for the upgrade) and eventually no plugin candidate found will also throw exception. (once we ignore scope). Hence upgradeApplication can create ApplicationUpdateDetail object based on different update failure cases that can happen and users can in theory do something about it. If upgradeApplication still throws exception that means its an internal error and it is not possible for users to do anything about it.

IMO, it will not be scalable to only throw exceptions and let caller create object on different kind of exception as we would than need to create different exception classes for each cases which might not be needed.

tl;dr, if upgradeApplication returns ApplicationUpdateDetail object, it is either success or a failure which is understandable by customers else throw internal error.